File: spec/python/specwrite/test_instance_std.py
import io
import unittest
from kaitaistruct import ConsistencyError, KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.instance_std import InstanceStd

class TestInstanceStd(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestInstanceStd, self).__init__(*args, **kwargs)
        self.struct_class = InstanceStd
        self.src_filename = 'src/str_encodings.bin'

    def test_check_shorter_header(self):
        r = InstanceStd()
        r.header = "1234"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: header,"):
            r._check()

    def test_check_longer_header(self):
        r = InstanceStd()
        r.header = "123456"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: header,"):
            r._check()

    def test_check_empty_header_via_dump(self):
        r = InstanceStd()
        r.header = ""
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: header,"):
            # calls all _check() methods
            CommonSpec.Base.dump_struct(r)

    def test_write(self):
        r = InstanceStd()
        r.header = u"Hello"

        # see <CommonSpec.Base>.test_read_write_roundtrip
        orig_dump = CommonSpec.Base.dump_struct(r)

        ks_io = KaitaiStream(io.BytesIO(bytearray(2 + 5)))
        r._write(ks_io)
        ks_io.seek(0)

        new_r = InstanceStd(ks_io)
        new_r._read()

        new_dump = CommonSpec.Base.dump_struct(new_r)

        self.assertEqual(new_dump, orig_dump)

File: spec/java/src/io/kaitai/struct/specwrite/TestInstanceStd.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.ConsistencyError;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.InstanceStd;
import org.testng.annotations.Test;
import static org.testng.Assert.*;
import java.lang.reflect.InvocationTargetException;

public class TestInstanceStd extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return InstanceStd.class;
    }

    @Override
    protected String getSrcFilename() {
        return "str_encodings.bin";
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: header,.*")
    public void testCheckShorterHeader() throws Exception {
        InstanceStd r = new InstanceStd();
        r.setHeader("1234");
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: header,.*")
    public void testCheckLongerHeader() throws Exception {
        InstanceStd r = new InstanceStd();
        r.setHeader("123456");
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: header,.*")
    public void testCheckEmptyHeaderViaDump() throws Exception {
        InstanceStd r = new InstanceStd();
        r.setHeader("");
        try {
            // calls all _check() methods
            dumpStruct(r);
        } catch (InvocationTargetException e) {
            if (e.getCause() instanceof Exception) {
                throw ((Exception) e.getCause());
            }
            throw e;
        }
    }

    @Test
    public void testWrite() throws Exception {
        InstanceStd r = new InstanceStd();
        r.setHeader("Hello");

        // see <CommonSpec>.testReadWriteRoundtrip
        Object origDump = dumpStruct(r);

        KaitaiStream io = new ByteBufferKaitaiStream(2 + 5);
        r._write(io);
        io.seek(0);

        InstanceStd newR = new InstanceStd(io);
        newR._read();

        Object newDump = dumpStruct(newR);

        assertEquals(newDump, origDump);
    }
}
